katello-attach-subscription

```diff
@@ -99,24 +135,769 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
+            # if empty, so zero results, this will be simply skipped
+            parsed_subscription.each do |subscription|
+              if @options[:debug]
+                puts "   DEBUG: subscription detail for subscription #{subscription['cp_id']}:"
+                p subscription
+              end
+              # if the api are old, subscription_id is set but cp_id is not. Standardizing.
+              if subscription.has_key?('subscription_id') and !(subscription.has_key?('cp_id')) and !(subscription['subscription_id'].is_a?(Integer))
+                subscription['cp_id'] = subscription.delete('subscription_id')
+              end
+              # The type is defined in yaml. Currently handling only the Hypervisor subs, derivated one should be not considered for hypervisors.
+              if subfiltertype and subscription.has_key?('type') and subscription['type'] != subfiltertype
+                if @options[:debug]
+                  puts "   Skipping '#{subscription['cp_id']}' as system type '#{subscription['type']}' is different from desired '#{subfiltertype}'"
+                end
+                # if the filter do not match, skip to next subscription
+                next
+              end
+              parsed_product_sub_array.push(subscription['cp_id'])
+            end
+            puts "    Subscription parsing results of search string '#{subscription_item}' is #{parsed_product_sub_array}"
+            # push new subs to array with new key for this product
+            sub_parsed[product] = parsed_product_sub_array
+            # and copy to the pointer for this product
+            # sub['sub'] are the requested subs
+            # sub['sub_parsed'] are the subs found that has to be associated
+            sub['sub_parsed']=sub_parsed
+          end
+        end
+      end
+    end
+    if @options[:debug]
+      puts " DEBUG: the full yaml subs output after modification"
+      p @yaml[:subs]
+    end
+  end
+
+  # if virt-who option is enabled we are going to execute the command virt-who --print to retrieve
+  # hypervisors data and so their number of socket
+  json_hash = {}
+  run_virtwho_print = false
+
+  if @options[:debug]
+    puts "DEBUG: virt-who enabled: #{@options[:virtwho]}"
+  end
+
+  if @options[:virtwho]
+    if @options[:debug]
+      puts "  DEBUG: Checking if virt-who is installed"
+    end
+    # virt-who cachefile contain the cached data of the json with the hypervisor data
+    # the desired pattern for the name of this cache file is virt-who-{ date in yyyymm }.cache_org{ organization id }
+    virtwho_cachefile = @options[:virtwhocachefile].to_s + "_org" + @options[:org].to_s
+    # firts of all, we have to check if the machine has virt-who installed
+    virtwho_location = `which virt-who`
+
+    virtwho_location = virtwho_location.chop!
+    # if which command exit and exit with code 0 then virt-who is installed in virtwho_location
+    if $?.exited? and $?.exitstatus.to_i == 0
```

Comment:
this check should happen directly after parsing `--args`, otherwise the use might end up waiting and then we error out late